rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,1025 @@
+FROM block_to_transactions
+GROUPBY block_to_transactions.state_hash
+HAVING ((block_to_transactions.sender IN $SENDERS) OR (block_to_transactions.receiver IN $RECEIVERS))
+) as group_by
+INNER JOIN
+blocks
+ON group_by.state_hash = blocks.state_hash
+WHERE group_by.time_seen > $TIME_SEEN
+ORDER BY time_seen
+LIMIT 10
+-- Would need to use a groupby query to make query the transactions as well
+```
+
+### Pooled_user_command
+
+```SQL
+-- querying specific receivers from the transaction pool
+SELECT * FROM
+transactions
+WHERE status = ADDED AND to = $RECEIVER
+```
+
+```SQL
+-- querying specific senders from the transaction pool
+SELECT * FROM
+transactions
+WHERE status = ADDED AND to = $RECEIVER
+```
+
+## Consistency Issues
+
+As mentioned in the [Running Bootstrap section](#bootstrap), we would not be able to fully determine the `consensus_status` of the block that were in the transition_frontier. Worse, we would not be able to determine which blocks became snarked.
+
+We can have a third-party archived node tell a client the consensus_state of the blocks. Another option that we have is that we can create a P2P RPC call where peers can give a client a Merkle list proof of a state_hash and all the state_hashes of the blocks on top of the state_hash. The Merkle list should be length `K`, the number of block confirmations to gaurantee finality with a very high probability. If the Merkle list proof is size `K`, then the block should be fully confirmed. Otherwise, the block failed.
+
+# Rationale and Alternatives
+
+- Makes it easy to perform difficult queries for our protocol
+- Makes the client code more module
+- The current implementation takes a lot of work and it is quite limited for a certain use case.
+- Other alternatives are discussed in the previous commits of this [RFC](https://github.com/CodaProtocol/coda/pull/2901)
+- In an ideal world where we have an infinite amount of time, we would implement our own native graph database leverage the `transition_frontier` as an in-memory cache. Namely, we would have all the blocks in the `transition_frontier` to be in the in-memory cache along with nodes that have a 1 to 2 degree of seperation
+
+# Prior Art
+
+Ethereum is representing their data as patricia tries and are storing them in LevelDB.
+
+Coinbase also uses MongoDB to store all of the data in a blockchain.
+
+# Unresolved questions
+
+- How can we extend this design so that a client has multiple devices (i.e. Desktop application, mobile, etc...).
+
+# Appendix
+
+The flexibility of this new design changed some parts of the current schema of the GraphqQL database. [../docs/res/graphql_diff.html](https://raw.githack.com/CodaProtocol/coda/rfc/storing-client-data/docs/res/graphql_diff.html) is a html file that shows the difference between these two graphql schemas. Here is the new proposed schema:
```

Comment:
As we discussed yesterday, please include a rough sketch of which parts of the existing graphql API would move to be backed by the DB and which parts would be handled by the daemon directly -- would be nice to cover all the queries (I'm assuming all subscriptions will go to sql server, and all mutations will go to daemon)

Given that we'll likely implement this incrementally by supporting our existing API first it's important to keep it as similar to that as possible for this reason.

In addition, you can include your proposed changes, but please be really specific about what you are changing and why -- and also include annotations on the queries for whether they are backed by daemon or database.